File: apps/ios/SayDo/NativeBridge.swift
import Foundation
import Combine
import WebKit

struct NativeFocus: Identifiable, Equatable {
    let id: String
    let title: String
    let updatedAt: String
}

enum NativeReplyOrigin: String {
    case assistantReply = "assistant_reply"
    case system
    case confirmation
    case onboarding
}

struct NativeReply: Equatable {
    let sessionId: String
    let turnId: String
    let sentenceId: String
    let text: String
    let origin: NativeReplyOrigin
}

struct NativeTranscriptSubmission {
    let requestId: String
    let captureId: String
    let text: String
    let focus: NativeFocus?
    let action: String

    var javaScriptArguments: [String: Any] {
        var input: [String: Any] = [
            "requestId": requestId,
            "captureId": captureId,
            "text": text,
            "action": action,
        ]
        if let focus {
            input["focusId"] = focus.id
            input["focusTitle"] = FocusTitleSanitizer.sanitize(focus.title)
        }
        return ["input": input]
    }
}

enum NativeSubmissionStatus: String {
    case queuedToSocket = "queued_to_socket"
    case drafted
    case rejected
}

struct NativeSubmissionResult {
    let status: NativeSubmissionStatus
    let reason: String?
}

enum NativeDaemonStatus: String {
    case connecting
    case online
    case offline
}

struct NativeBridgeOrigin: Equatable {
    let scheme: String
    let host: String
    let port: Int

    init?(url: URL) {
        guard let scheme = url.scheme?.lowercased(), let host = url.host?.lowercased() else { return nil }
        self.scheme = scheme
        self.host = host
        port = url.port ?? Self.defaultPort(for: scheme)
    }

    func matches(scheme candidateScheme: String, host candidateHost: String, port candidatePort: Int) -> Bool {
        scheme == candidateScheme.lowercased()
            && host == candidateHost.lowercased()
            && port == candidatePort
    }

    func matches(url: URL?) -> Bool {
        guard let url, let candidate = NativeBridgeOrigin(url: url) else { return false }
        return candidate == self
    }

    private static func defaultPort(for scheme: String) -> Int {
        switch scheme {
        case "http": 80
        case "https": 443
        default: -1
        }
    }
}

@MainActor
final class NativeBridgeController: NSObject, ObservableObject, WKScriptMessageHandler {
    nonisolated static let version = 1
    nonisolated static let handlerName = "saydoNative"

    @Published private(set) var pageReady = false
    @Published private(set) var focuses: [NativeFocus] = []
    @Published private(set) var daemonStatus: NativeDaemonStatus = .connecting

    var onReply: ((NativeReply) -> Void)?
    private weak var webView: WKWebView?
    private var pageGeneration = 0
    private var activeSessionId: String?
    private var expectedOrigin: NativeBridgeOrigin?
    private var expectedPageNonce: String?

    func attach(webView: WKWebView) {
        self.webView = webView
        invalidatePage()
    }

    func bind(profileURL: URL) {
        let origin = NativeBridgeOrigin(url: profileURL)
        guard expectedOrigin != origin else { return }
        expectedOrigin = origin
        invalidatePage()
    }

    func beginNavigation(pageNonce: String) {
        expectedPageNonce = pageNonce
        invalidatePage(keepingNonce: true)
    }

    func invalidatePage(keepingNonce: Bool = false) {
        pageGeneration += 1
        pageReady = false
        activeSessionId = nil
        focuses = []
        daemonStatus = .connecting
        if !keepingNonce {
            expectedPageNonce = nil
        }
    }

    func allowsProfileURL(_ url: URL?) -> Bool {
        expectedOrigin?.matches(url: url) == true
    }

    func userContentController(_ userContentController: WKUserContentController, didReceive message: WKScriptMessage) {
        guard message.name == Self.handlerName,
              message.frameInfo.isMainFrame,
              let webView,
              message.webView === webView,
              let expectedOrigin,
              expectedOrigin.matches(
                scheme: message.frameInfo.securityOrigin.protocol,
                host: message.frameInfo.securityOrigin.host,
                port: message.frameInfo.securityOrigin.port
              ) else {
            return
        }
        guard let body = message.body as? [String: Any],
              (body["version"] as? NSNumber)?.intValue == Self.version,
              body["pageNonce"] as? String == expectedPageNonce,
              let type = body["type"] as? String else {
            return
        }
        switch type {
        case "page-ready":
            guard let sessionId = body["sessionId"] as? String, !sessionId.isEmpty else { return }
            activeSessionId = sessionId
            pageReady = true
        case "focuses":
            guard pageReady else { return }
            focuses = parseFocuses(body["focuses"])
        case "status":
            guard pageReady else { return }
            guard let raw = body["status"] as? String, let status = NativeDaemonStatus(rawValue: raw) else { return }
            daemonStatus = status
        case "reply":
            guard pageReady else { return }
            guard let reply = parseReply(body["reply"]), reply.sessionId == activeSessionId else { return }
            onReply?(reply)
        default:
            return
        }
    }

    func submit(_ submission: NativeTranscriptSubmission) async -> NativeSubmissionResult {
        guard pageReady, let webView, allowsProfileURL(webView.url) else {
            return NativeSubmissionResult(status: .rejected, reason: "bridge_inactive")
        }
        let expectedPageGeneration = pageGeneration
        let body = """
        const bridge = window.SayDoNativeBridge;
        if (!bridge || bridge.version !== 1) {
          return { status: "rejected", reason: "bridge_inactive" };
        }
        return await bridge.submitNativeTranscript(input);
        """
        do {
            let raw = try await webView.callAsyncJavaScript(
                body,
                arguments: submission.javaScriptArguments,
                in: nil,
                contentWorld: .page
            )
            guard pageReady,
                  pageGeneration == expectedPageGeneration,
                  self.webView === webView,
                  allowsProfileURL(webView.url) else {
                return NativeSubmissionResult(status: .rejected, reason: "bridge_inactive")
            }
            guard let result = raw as? [String: Any],
                  let statusRaw = result["status"] as? String,
                  let status = NativeSubmissionStatus(rawValue: statusRaw),
                  result["requestId"] as? String == submission.requestId,
                  result["captureId"] as? String == submission.captureId else {
                return NativeSubmissionResult(status: .rejected, reason: "invalid_bridge_response")
            }
            return NativeSubmissionResult(status: status, reason: result["reason"] as? String)
        } catch {
            return NativeSubmissionResult(status: .rejected, reason: "bridge_call_failed")
        }
    }

    private func parseFocuses(_ raw: Any?) -> [NativeFocus] {
        guard let rows = raw as? [[String: Any]] else { return [] }
        return rows.prefix(3).compactMap { row in
            guard let id = row["id"] as? String,
                  let title = row["title"] as? String,
                  let updatedAt = row["updatedAt"] as? String else {
                return nil
            }
            return NativeFocus(id: id, title: FocusTitleSanitizer.sanitize(title), updatedAt: updatedAt)
        }
    }

    private func parseReply(_ raw: Any?) -> NativeReply? {
        guard let value = raw as? [String: Any],
              let sessionId = value["sessionId"] as? String,
              let turnId = value["turnId"] as? String,
              let sentenceId = value["sentenceId"] as? String,
              let text = value["text"] as? String,
              let originRaw = value["origin"] as? String,
              let origin = NativeReplyOrigin(rawValue: originRaw) else {
            return nil
        }
        return NativeReply(
            sessionId: sessionId,
            turnId: turnId,
            sentenceId: sentenceId,
            text: text,
            origin: origin
        )
    }
}

File: packages/console/src/mobile/nativeBridge.ts
import type { FocusRow } from "./types";

export const NATIVE_BRIDGE_VERSION = 1 as const;
export const NATIVE_BRIDGE_HANDLER = "saydoNative";

export interface NativeFocus {
  id: string;
  title: string;
  updatedAt: string;
}

export interface NativeReply {
  sessionId: string;
  turnId: string;
  sentenceId: string;
  text: string;
  origin: "assistant_reply" | "system" | "confirmation" | "onboarding";
}

export interface NativeTranscriptRequest {
  requestId: string;
  captureId: string;
  text: string;
  focusId?: string;
  focusTitle?: string;
  action: "send" | "draft";
}

export type NativeTranscriptResult =
  | { status: "queued_to_socket"; requestId: string; captureId: string }
  | { status: "drafted"; requestId: string; captureId: string }
  | {
      status: "rejected";
      requestId: string;
      captureId: string;
      reason: "invalid_request" | "bridge_inactive" | "offline" | "busy" | "draft_conflict" | "focus_mismatch";
    };

interface NativeMessageHandler {
  postMessage(message: unknown): void;
}

export interface NativeBridgeWindow {
  __saydoNativePageNonce?: string;
  webkit?: { messageHandlers?: Record<string, NativeMessageHandler | undefined> };
  SayDoNativeBridge?: {
    version: typeof NATIVE_BRIDGE_VERSION;
    submitNativeTranscript(input: unknown): Promise<NativeTranscriptResult>;
  };
}

export interface NativeBridgeConnection {
  publishFocuses(focuses: NativeFocus[]): void;
  publishStatus(status: "connecting" | "online" | "offline"): void;
  forwardReply(reply: NativeReply): void;
  dispose(): void;
}

function rejectInvalid(input: unknown): NativeTranscriptResult {
  const value = typeof input === "object" && input !== null ? input as Record<string, unknown> : {};
  return {
    status: "rejected",
    requestId: typeof value["requestId"] === "string" ? value["requestId"] : "",
    captureId: typeof value["captureId"] === "string" ? value["captureId"] : "",
    reason: "invalid_request"
  };
}

export function parseNativeTranscriptRequest(input: unknown): NativeTranscriptRequest | null {
  if (typeof input !== "object" || input === null || Array.isArray(input)) return null;
  const value = input as Record<string, unknown>;
  const allowed = new Set(["requestId", "captureId", "text", "focusId", "focusTitle", "action"]);
  if (Object.keys(value).some((key) => !allowed.has(key))) return null;
  const hasFocusId = typeof value["focusId"] === "string" && value["focusId"].trim() !== "";
  const hasFocusTitle = typeof value["focusTitle"] === "string" && value["focusTitle"].trim() !== "";
  if (
    typeof value["requestId"] !== "string" || value["requestId"].trim() === "" ||
    typeof value["captureId"] !== "string" || value["captureId"].trim() === "" ||
    typeof value["text"] !== "string" || value["text"].trim() === "" ||
    (value["action"] !== "send" && value["action"] !== "draft") ||
    (value["focusId"] !== undefined && typeof value["focusId"] !== "string") ||
    (value["focusTitle"] !== undefined && typeof value["focusTitle"] !== "string") ||
    hasFocusId !== hasFocusTitle
  ) return null;
  return {
    requestId: value["requestId"],
    captureId: value["captureId"],
    text: value["text"],
    action: value["action"],
    ...(typeof value["focusId"] === "string" ? { focusId: value["focusId"] } : {}),
    ...(typeof value["focusTitle"] === "string" ? { focusTitle: value["focusTitle"] } : {})
  };
}

export function sanitizeNativeFocusTitle(title: string): string {
  return Array.from(title.replace(/\]/gu, "").replace(/[\r\n]+/gu, " ").replace(/\s+/gu, " ").trim())
    .slice(0, 64)
    .join("");
}

export function recentNativeFocuses(rows: FocusRow[] | null): NativeFocus[] {
  if (!rows) return [];
  return rows
    .filter((row) => row.lifecycle === "captured" || row.lifecycle === "active" || row.lifecycle === "dormant")
    .sort((a, b) => b.updatedAt.localeCompare(a.updatedAt) || a.id.localeCompare(b.id))
    .slice(0, 3)
    .map((row) => ({ id: row.id, title: sanitizeNativeFocusTitle(row.title), updatedAt: row.updatedAt }));
}

export function composeNativeTranscript(request: NativeTranscriptRequest, focuses: NativeFocus[]): string | null {
  const text = request.text.trim();
  if (!request.focusId && !request.focusTitle) return text;
  const focus = focuses.find((item) => item.id === request.focusId);
  if (!focus || focus.title !== sanitizeNativeFocusTitle(request.focusTitle ?? "")) return null;
  return `[关于:${focus.title}] ${text}`;
}

export function installNativeBridge(
  target: NativeBridgeWindow,
  sessionId: string,
  submit: (request: NativeTranscriptRequest) => Promise<NativeTranscriptResult> | NativeTranscriptResult
): NativeBridgeConnection | null {
  const handler = target.webkit?.messageHandlers?.[NATIVE_BRIDGE_HANDLER];
  const pageNonce = target.__saydoNativePageNonce;
  if (!handler || typeof pageNonce !== "string" || pageNonce === "") return null;
  let active = true;
  const post = (type: string, payload: Record<string, unknown> = {}) => {
    if (active) handler.postMessage({ version: NATIVE_BRIDGE_VERSION, pageNonce, type, ...payload });
  };
  target.SayDoNativeBridge = {
    version: NATIVE_BRIDGE_VERSION,
    async submitNativeTranscript(input: unknown) {
      if (!active) {
        const invalid = rejectInvalid(input);
        return { ...invalid, reason: "bridge_inactive" };
      }
      const request = parseNativeTranscriptRequest(input);
      return request ? await submit(request) : rejectInvalid(input);
    }
  };
  post("page-ready", { sessionId });
  return {
    publishFocuses: (focuses) => post("focuses", { focuses }),
    publishStatus: (status) => post("status", { status }),
    forwardReply: (reply) => post("reply", { reply }),
    dispose() {
      active = false;
      if (target.SayDoNativeBridge?.version === NATIVE_BRIDGE_VERSION) delete target.SayDoNativeBridge;
    }
  };
}
